light-prover/prover/update_circuit.go

```diff
@@ -0,0 +1,296 @@
+package prover
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/constraint"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+	"github.com/reilabs/gnark-lean-extractor/v2/abstractor"
+)
+
+type BatchUpdateCircuit struct {
+	PreRoot    frontend.Variable `gnark:",public"`
+	PostRoot   frontend.Variable `gnark:",public"`
+	StartIndex frontend.Variable `gnark:"input"`
+
+	OldLeaves    []frontend.Variable   `gnark:"input"`
+	NewLeaves    []frontend.Variable   `gnark:"input"`
```

Comment:
Note: `NewLeaves` will always be 0 for state Merkle tree updates. I can see merit in making the circuit generic. However, we should make sure that dynamic `NewLeaves` don't come at a performance penalty.
Have you benchmarked dynamic `NewLeaves` vs hardcoding an update with '0' in circuit?